### ISSUE TITLE
Device binding according to process

### ISSUE THREAD

**Issue by LudovicEnault** [open-mpi/ompi#7857]
This is not an issue, it's a feature I would like to add first myself in OpenMPI, and then it might interest the rest of the community.
For a start I am wondering what is the best way & location to add such modification in OpenMPI: adapt device(s) used by each process depending on general setting.
E.G:
MPI task | CPU binding | socket | Device binding
---------- | -- | -- | ----
0  | 0-23 | 0 | mlx5_0:1
1 | 24-47 | 1 | mlx5_1:1

(I am also wondering if this is not possible to do that at the Job scheduler level).

## Background information
I am trying to run on a four socket system with two IB connection.
I was able to demonstrate that, in my case, "binding" to a specific device would decrease communication time (latency) by limiting the umber of CPU-CPU hops.
The demonstration was made by using a script that would change the environment before runing the mpi process:
```mpirun ./ucs_device.sh mpi_application application_options``` 
In the above example, the script would do:
task 0: `export UCX_NET_DEVICES=mlx5_0:1` 
task 1: `export UCX_NET_DEVICES=mlx5_1:1` 
and voilà, 20% better performance.

Any suggestion where to start ?

### What version of Open MPI are you using? (e.g., v3.0.5, v4.0.2, git branch name and hash, etc.)
I have currently based my tests on OpenMPI 4.0.3

### Describe how Open MPI was installed (e.g., from a source/distribution tarball, from a git clone, from an operating system distribution package, etc.)
Package was installed with UCX and PMIX (without verbs)

### If you are building/installing from a git clone, please copy-n-paste the output from `git submodule status`.
openmpi-4.0.3.tar.gz official tarball

### Please describe the system on which you are running

* Operating system/version: RedHat 8.0
* Computer hardware: four socket x86 System
* Network type: Infiniband EDR, dual attachement

-----------------------------

## Details of the problem
Described above.

**Comment by yosefe** [comment]
@LudovicEnault UCX by default picks the device which is closes to the current thread's CPU.
So using --bind-to option and UCX_NET_DEVICES=mlx5_0:1,mlx5_1:1 (or not passing UCX_NET_DEVICES at all) should ensure optimal usage

**Comment by LudovicEnault** [comment]
Thanks for your feedback.
Maybe it should but it does not, or the one that I specify is better.
Do you mean this should be something to check in UCX rather than OpenMPI ?

**Comment by yosefe** [comment]
@LudovicEnault yes, can you please specify exactly waht you are running (versions, command line), maybe also hwloc-ls output) and how do you see that the behavior is not as expected (e.g a sub-optimal device is used)?

**Comment by LudovicEnault** [comment]
I am using OpenMPI 4.0.3, UCX 1.8.0, PMIX 3.1.5 KNEM coming with Mellanox drivers.
The commands are:
1. With "default binding":
```
export export UCX_NET_DEVICES="mlx5_0:1,mlx5_2:1"
srun --mpi=pmix_v3 ./application
```
2. with manual device assignement:
```
srun --mpi=pmix_v3 ./ucx_net_devices.sh ./application
```
ucx_net_devices.sh will export UCX_NET_DEVICES=mlx5_0:1 for process running on socket 0 and 2 and mlx5_2:1 for process running on socket 1 and 3.
In the four socket topology CPU 2 is directly connected to CPU0 and CPU3 to CPU1:
```
CPU2 --- CPU0 --- mlx5_0:1
  |       | 
CPU3 --- CPU1 --- mlx5_2:1
```
The outcome is with manual binding it's 10% faster (reproducible):
binding | elapsed time [s]
-- | --
auto | 467
manual | 419

[hwloc-ls.txt](https://github.com/open-mpi/ompi/files/4815942/hwloc-ls.txt)

**Comment by yosefe** [comment]
@LudovicEnault thanks, so maybe the the problem is relating to making the right choice for CPU2 and CPU3.
1. can you pls attach the output of `numactl -H` and `lscpu`?
2. how was UCX ./configure'd? by default it tries to take libnuma, unless `--disable-numa` is passed?
3. can you pls attach the script ./ucx_net_devices.sh for reference of the right way to select the device?

**Comment by LudovicEnault** [comment]
Hello,
1. output of numactl -h and lspcu attached
2. I used the following option to compile UCX (those used by Mellanox but with logging capabilities): `--build=x86_64-redhat-linux-gnu --host=x86_64-redhat-linux-gnu --disable-dependency-tracking --disable-optimizations --enable-logging --disable-debug --disable-assertions --disable-params-check --enable-cma --without-cuda --without-gdrcopy --with-verbs --with-knem --with-rdmacm --without-rocm --without-xpmem --without-ugni --enable-mt`
In the log file I can see libnuma has been found, and the binaries are built with libnuma:

```
[root@a2n01 1.8.0]# ldd bin/ucx_info  | grep numa
        libnuma.so.1 => /lib64/libnuma.so.1 (0x00007f1d99e03000)
```

3. I have attached ucx_net_devices.sh script

4. I checked the dynamic device assignment with `export UCX_LOG_LEVEL=data`.
Even if UCX_MAX_RNDV_RAILS is set to 1 (as described here: http://openucx.github.io/ucx/faq.html), the processes on socket 1 are bound to two devices:
```
[enaultl@login1 work]$ ./check-ucx-info.sh ucx_default/job.log
rank  socket  device
0     0       mlx5_0:1
1     0       mlx5_0:1
2     0       mlx5_0:1
3     1       mlx5_0:1  mlx5_2:1
4     1       mlx5_0:1  mlx5_2:1
5     1       mlx5_0:1  mlx5_2:1
6     2       mlx5_0:1
7     2       mlx5_0:1
8     2       mlx5_0:1
9     3       mlx5_2:1
10    3       mlx5_2:1
11    3       mlx5_2:1
[enaultl@login1 work]$ ./check-ucx-info.sh ucx_net_devices.sh/job.log
rank  socket  device
0     0       mlx5_0:1
1     0       mlx5_0:1
2     0       mlx5_0:1
3     1       mlx5_2:1
4     1       mlx5_2:1
5     1       mlx5_2:1
6     2       mlx5_0:1
7     2       mlx5_0:1
8     2       mlx5_0:1
9     3       mlx5_2:1
10    3       mlx5_2:1
11    3       mlx5_2:1
```
[numactl-H.txt](https://github.com/open-mpi/ompi/files/4836112/numactl-H.txt)
[lscpu.txt](https://github.com/open-mpi/ompi/files/4836113/lscpu.txt)
[ucx_net_devices.sh.txt](https://github.com/open-mpi/ompi/files/4836131/ucx_net_devices.sh.txt)

**Comment by yosefe** [comment]
@LudovicEnault thanks for the info, so the problem with UCX is selecting the right device for socket1: it selects both mlx5_0:1 and mlx5_2:1 , but it should select only mlx5_2:1
Can you pls provide also:
1. The script check-ucx-info.sh
2. output of:
```bash
for x in /sys/class/infiniband/*
do
    echo $x $(cat $x/device/local_cpulist) $(cat $x/device/local_cpus) $(cat $x/device/numa_node)
done
```

**Comment by LudovicEnault** [comment]
Here you go:
1. [check-ucx-info.sh.txt](https://github.com/open-mpi/ompi/files/4845966/check-ucx-info.sh.txt)
2. [mlx5-device-info.txt](https://github.com/open-mpi/ompi/files/4845968/mlx5-device-info.txt) (only mlx5_0 and mlx5_2 have cables connected).
```
/sys/class/infiniband/mlx5_0  0-23,96-119    00000000,00000000,00ffffff,00000000,00000000,00ffffff  0
/sys/class/infiniband/mlx5_1  0-23,96-119    00000000,00000000,00ffffff,00000000,00000000,00ffffff  0
/sys/class/infiniband/mlx5_2  24-47,120-143  00000000,0000ffff,ff000000,00000000,0000ffff,ff000000  1
/sys/class/infiniband/mlx5_3  24-47,120-143  00000000,0000ffff,ff000000,00000000,0000ffff,ff000000  1
```

**Comment by yosefe** [comment]
@LudovicEnault seems like different devices are selected for different endpoints (for different connections).
1. On my setup, i see potentially similar problem, which is resolved by adding `-mca coll ^hcoll` to mpirun. Can you pls check this flag as well?
1. can you pls upload ucx_default/job.log?

**Comment by LudovicEnault** [comment]
1. I am checking
2. [ucx_default-job.log.txt](https://github.com/open-mpi/ompi/files/4852325/ucx_default-job.log.txt), I made some cleaning (removing "UCX  WARN  object ... was not returned to mpool ucp_requests" and stuff like that)
